core/toolbox/toolbox_item.js

```diff
@@ -0,0 +1,224 @@
+ *     defining the toolbox item.
+ * @param {!Blockly.IToolbox} toolbox The toolbox that holds the toolbox item.
+ * @constructor
+ * @extends {Blockly.ToolboxItem}
+ */
+Blockly.SelectableToolboxItem = function(toolboxItemDef, toolbox) {
+  Blockly.SelectableToolboxItem.superClass_.constructor.call(
+      this, toolboxItemDef, toolbox);
+};
+Blockly.utils.object.inherits(Blockly.SelectableToolboxItem,
+    Blockly.ToolboxItem);
+
+/**
+ * @override
+ */
+Blockly.SelectableToolboxItem.prototype.isSelectable = function() {
+  return true;
+};
+
+/**
+ * Gets the name of the toolbox item. Used for emitting events.
+ * @return {string} The name of the toolbox item.
+ * @public
+ */
+Blockly.SelectableToolboxItem.prototype.getName = function() {
+  return '';
+};
+
+/**
+ * Gets the contents of the toolbox item. These are items that are meant to be
+ * displayed in the flyout.
+ * @return {!Blockly.utils.toolbox.FlyoutItemJsonArray|string} The definition
+ *     of items to be displayed in the flyout.
+ * @public
+ */
+Blockly.SelectableToolboxItem.prototype.getContents = function() {
+  return [];
+};
+
+/**
+ * Set the current toolbox item as selected. No-op by default.
+ * @param {boolean} _isSelected True if this category is selected, false
+ *     otherwise.
+ * @public
+ */
+Blockly.SelectableToolboxItem.prototype.setSelected = function(_isSelected) {
+};
+
+/**
+ * Event listener for when the toolbox item is clicked.
+ * @param {!Event} _e Click event to handle.
+ * @public
+ */
+Blockly.SelectableToolboxItem.prototype.onClick = function(_e) {};
+
+/**
+ * Class for an item in the toolbox that is collapsible.
+ * @param {!Blockly.utils.toolbox.ToolboxItemJson} toolboxItemDef The JSON
+ *     defining the toolbox item.
+ * @param {!Blockly.IToolbox} toolbox The toolbox that holds the toolbox item.
+ * @constructor
+ * @extends {Blockly.SelectableToolboxItem}
+ */
+Blockly.CollapsibleToolboxItem = function(toolboxItemDef, toolbox) {
+  Blockly.CollapsibleToolboxItem.superClass_.constructor.call(
+      this, toolboxItemDef, toolbox);
+
+};
+Blockly.utils.object.inherits(Blockly.CollapsibleToolboxItem,
+    Blockly.SelectableToolboxItem);
+
+/**
+ * @override
+ */
+Blockly.CollapsibleToolboxItem.prototype.isCollapsible = function() {
+  return true;
+};
+
+/**
+ * Gets any children toolbox items. (ex. Gets the sub categories)
+ * @return {!Array<!Blockly.ToolboxItem>} The child toolbox items.
+ */
+Blockly.CollapsibleToolboxItem.prototype.getChildToolboxItems = function() {
+  return [];
+};
+
+/**
+ * Whether the toolbox item is expanded to show it's child subcategories.
+ * @return {boolean} True if the toolbox item shows it's children, false if it
+ *     is collapsed.
+ * @public
+ */
+Blockly.CollapsibleToolboxItem.prototype.isExpanded = function() {
+  return false;
+};
+
+/**
+ * Toggles whether or not the toolbox item is expanded.
+ * @public
+ */
+Blockly.CollapsibleToolboxItem.prototype.toggleExpanded = function() {};
```

Comment:
An alternate option is to make collapsible an interface, and make two versions of Category: one that extends selectable, and one that extends the base Category and implements `ICollapsible`.

Let's discuss offline.

Reply:
Interface was created and Category was separated into CollapsibleToolboxCategory and ToolboxCategory: #4244